pkg/api/node_api.go

```diff
@@ -916,6 +918,141 @@ func (a *NodeApi) snapshotStateHash(w http.ResponseWriter, r *http.Request) erro
 	return nil
 }
 
+type GeneratorInfo struct {
+	Address       string `json:"address"`
+	Balance       uint64 `json:"balance"`
+	TransactionID string `json:"transactionID"`
+}
+
+func (a *NodeApi) GeneratorsAt(w http.ResponseWriter, r *http.Request) error {
+	heightStr := chi.URLParam(r, "height")
+	height, err := strconv.ParseUint(heightStr, 10, 64)
+	if err != nil {
+		return errors.Wrap(err, "invalid height")
+	}
+
+	activationHeight, err := a.state.ActivationHeight(int16(settings.DeterministicFinality))
+	if err != nil {
+		return fmt.Errorf("failed to get DeterministicFinality activation height: %w", err)
+	}
+
+	periodStart, err := state.CurrentGenerationPeriodStart(activationHeight, height,
+		a.app.settings.GenerationPeriod)
+	if err != nil {
+		return err
+	}
+
+	var generatorsInfo []GeneratorInfo
+
+	generatorAddresses, err := a.state.CommittedGenerators(periodStart)
+	if err != nil {
+		return err
+	}
+	for _, generatorAddress := range generatorAddresses {
+		endorserRecipient := proto.NewRecipientFromAddress(generatorAddress)
+		balance, pullErr := a.state.GeneratingBalance(endorserRecipient, height)
+		if pullErr != nil {
+			return pullErr
+		}
+		generatorsInfo = append(generatorsInfo, GeneratorInfo{
+			Address:       generatorAddress.String(),
+			Balance:       balance,
+			TransactionID: "", // TODO should be somehow found.
+		})
+	}
+	return trySendJSON(w, generatorsInfo)
+}
+
+func (a *NodeApi) FinalizedHeight(w http.ResponseWriter, _ *http.Request) error {
+	h, err := a.state.LastFinalizedHeight()
+	if err != nil {
+		return err
+	}
+	return trySendJSON(w, map[string]uint64{"height": h})
+}
+
+func (a *NodeApi) FinalizedHeader(w http.ResponseWriter, _ *http.Request) error {
+	blockHeader, err := a.app.state.LastFinalizedBlock()
+	if err != nil {
+		return err
+	}
```

Comment:
No handling of defined finalization errors: `ErrNoFinalization` and `ErrNoFinalizationHistory`. (`500` error)